File: problema2.js
// Lista original de diccionarios
const listaOriginal = [
    { 'make': 'Nokia', 'model': 216, 'color': 'Black' },
    { 'make': 'Mi Max', 'model': 2, 'color': 'Gold' },
    { 'make': 'Samsung', 'model': 7, 'color': 'Blue' }
];


const ordernarKey =(lista, key) => {
    return [...lista].sort((a,b)=> {
        if(a[key] < b[key]) return -1
        if(a[key] > b[key]) return 1
        return 0
    })
}

const listaOrdenada = ordernarKey(listaOriginal, 'model')
// Mostrar resultados
console.log('Lista original:');
console.log(listaOriginal);

console.log('\nLista ordenada por "model":');
console.log(listaOrdenada);

File: problema2.hs


-- Definir el tipo de dato para los teléfonos
data Phone = Phone { make :: String
                   , model :: Int
                   , color :: String
                   } deriving (Show)


--Lista original

listaOriginal::[Phone]
listaOriginal = [ Phone "Nokia" 216 "Black"
                , Phone "Mi Max" 2 "Gold"
                , Phone "Samsung" 7 "Blue"
                ]


--Función Lambda para ordenar usando el algoritmo de quicksort
ordenarPorModelo :: [Phone] -> [Phone]
ordenarPorModelo [] = []
ordenarPorModelo(pivote:resto) =
    let menores = filter (\phone -> model phone < model pivote) resto
        mayores = filter (\phone -> model phone >= model pivote) resto
    in ordenarPorModelo menores ++ [pivote] ++ ordenarPorModelo mayores
--Lista ordenada 
listaOrdenada ::[Phone]
listaOrdenada = ordenarPorModelo listaOriginal

main :: IO ()
main = do
    putStrLn "Lista original:"
    mapM_ print listaOriginal
    putStrLn "\nLista ordenada por 'model':"
    mapM_ print listaOrdenada